core/server/master/src/main/java/alluxio/master/meta/MetaMasterSync.java

```diff
@@ -143,6 +131,29 @@ private void handleCommand(MetaCommand cmd) throws IOException {
     }
   }
 
+  /**
+   * Sets the master id and registers with the Alluxio leader master.
+   */
+  private void setIdAndRegister() throws IOException {
```

Comment:
not thread safe currently. I assume you want to have thread-safey there and thus use `AtomicReference`.
However, the check  and set is separated without atomicity. I would understand better our intention here.

Reply:
The metaMasterSync would only be executed by one thread? 
And we normally only run into the if(mMasterId.get() == -1L）for once which is at the beginning. AtomicReference here is more for set and get value.

In addition, if we reformat the ```setId``` function to ```getId ```function, which we need a local ```AtomicReference masterId``` to return, and set it in the ```mMasterId.compareAndSet(-1L, getId())```, which looks strange.

Reply:
The best way is what the BlockMasterSync does, we getWorkerId in the defaultBlockWorker first and then pass it into BlockMasterSync.

However, if we did the same thing in MetaMasterSync, all masters will stuck in getMasterId because no leader master is available at the beginning. This is the same reason why we cannot getMasterId in the constructuor of MetaMasterSync.

```
      if (mMasterId.get() == UNINITIALIZED_MASTER_ID) {
        setIdAndRegister();
      }
```
This code are actually initializer, get the master id and register with master in the first heartbeat. The following heartbeat actually only checks if the id is still valid, and if not, set id and register again.

Reply:
the current code is good since there is no thread contention.